plugins/ltac/rewrite.ml

```diff
@@ -1468,6 +1467,9 @@ let cl_rewrite_clause_aux ?(abs=None) strat env avoid sigma concl is_hyp : resul
       let (_, cstrs) = res.rew_evars in
       let evars' = solve_constraints env res.rew_evars in
       let newt = Reductionops.nf_evar evars' res.rew_to in
+      (* Substitute solved constraint evars in newly created goals so that we can remove them *)
+      let evars' = Evd.raw_map_undefined (fun ev info ->
```

Comment:
It's potentially quite expensive to fold over all undefined evars. I don't know whether the call to solve_constraints above is already O(n), but we should check for performance. Also, uses of `Evd.mem` often indicate you're doing something fishy with evars. Setoid rewriting is indeed performing evar surgery right there, as `Evd.remove` is a very unsafe piece of code that should ultimately disappear. Maybe it's the right time to rethink a bit what this is supposed to do?

Reply:
I agree all this looks quite unsafe and crazy. But I´m not enough of an expert to weigh in on a proper fix.

Regarding the complexity: I think it is okay to do this here. There is already a map over the undefined evars in the same evar_map a little bit later in the code path: https://github.com/coq/coq/blob/9fe4108289f584461b5dc7af08095d6279a222af/plugins/ltac/rewrite.ml#L1552-L1554

Reply:
The other easy approach to fix this is to simply remove all the `nf_evar` calls in these functions, and then simply not remove the constraint evars. That sounds much safer and simpler to me. But I´m not sure what the performance impact of that would be...

Reply:
Let's launch a bench and see what happens. If no suspicious slowdown is observed we should merge this PR, as it fixes a bug. Redesigning setoid rewrite should be left to another time.

Reply:
I think this is pretty much a remnant of the old days, before the engine could handle evars correctly everywhere. So we could just leave the constraint evars where they are. If we find out it is necessary then I guess we should rather be calling a compaction/garbage collection on the evar_map instead.

Reply:
Do you propose I make that change in this PR or to leave it for some other time (both is fine with me)?

Reply:
I agree with Pierre-Marie, you can first get this bugfix merge and do it after. We'll have the most precise timing difference this way. I noticed in `coq-performance-tests` there's a big improvement in max resident memory but I'm not sure it really means anything.